keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1297 @@
+pkg/cri/annotations/annotations.go
+
+  // SandboxCPU annotations are based on the initial CPU configuration for the sandbox. This is calculated as the
+  // sum of container CPU resources, optionally provided by Kubelet (introduced  in 1.23) as part of the PodSandboxConfig
+  SandboxCPUPeriod = "io.kubernetes.cri.sandbox-cpu-period"
+  SandboxCPUQuota  = "io.kubernetes.cri.sandbox-cpu-quota"
+  SandboxCPUShares = "io.kubernetes.cri.sandbox-cpu-shares"
+
+  // SandboxMemory is the initial amount of memory associated with this sandbox. This is calculated as the sum
+  // of container memory, optionally provided by Kubelet (introduced in 1.23) as part of the PodSandboxConfig.
+  SandboxMem = "io.kubernetes.cri.sandbox-memory"
+```
+
+However, the original details of
+the resource spec are lost as they get translated (within kubelet) to
+platform-specific (i.e. Linux or Windows) resource controller parameters like
+cpu shares, memory limits etc. Non-native resources such as extended resources
+and the device plugin resources completely invisible to the CRI runtime. However,
+[OCI hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md),
+[runC](https://github.com/opencontainers/runc) wrappers,
+[NRI](https://github.com/containerd/nri) plugins or in some cases even
+applications themselves would benefit on seeing the original resource requests
+and limits e.g. for doing customized resource optimization.
+
+Extending the CRI API to communicate all resources already at sandbox creation
+and pass down resource requests and limits (of native and extended resources)
+would provide a comprehensive and early-enough view of the resource usage of
+all containers of the pod, allowing improved resource allocation without
+breaking any existing use cases.
+
+### Goals
+
+- make the information about all required resources (e.g. native and extended
+  resources, devices, mounts, CDI devices) of a Pod available to the CRI at
+  sandbox creation time
+- make container resource spec transparently visible to CRI (the container
+  runtime)
+
+### Non-Goals
+
+- change kubelet resource management
+- change existing behavior of CRI
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a VM-based container runtime developer, I want to allocate/expose enough
+RAM, hugepages, hot- or cold-pluggable PCI(e) ports, protected memory sections
+and other resources for the VM to ensure that all containers in the pod are
+guaranteed to get the resources they require.
+
+#### Story 2
+
+As a developer of non-runc / non-Linux CRI runtime, I want to know detailed
+container resource requests to be able to make correct resource allocation for
+the applications. I cannot rely on cgroup parameters on this but need to know
+what the user requested to fairly allocate resources between applications.
+
+#### Story 3
+
+As a cluster administrator, I want to install an NRI plugin that does
+customized resource handling. I run kubelet with CPU manager and memory manager
```

Comment:
A similar case but something that could be a story of its own stresses the goal for having better visibility to extended resources:

"As a cluster administrator, I want to install an NRI plugin that does limits enforcement for extended resources that kubelet does not know about (e.g., `misc` cgroup settings). These extended resources are consumed either via the pod spec or indirectly via Pod `overhead`".

Reply:
Thx @mythi added this as a separate user story. WDYT?